hugr-core/src/export.rs

```diff
@@ -822,54 +818,25 @@ impl<'a> Context<'a> {
         }
     }
 
-    pub fn export_ext_set(&mut self, t: &ExtensionSet) -> model::TermId {
-        // Extension sets with variables are encoded using a hack: a variable in the
-        // extension set is represented by converting its index into a string.
-        // Until we have a better representation for extension sets, we therefore
-        // need to try and parse each extension as a number to determine if it is
-        // a variable or an extension.
-
-        // NOTE: This overprovisions the capacity since some of the entries of the row
-        // may be variables. Since we panic when there is more than one variable, this
-        // may at most waste one slot. That is way better than having to allocate
-        // a temporary vector.
-        //
-        // Also `ExtensionSet` has no way of reporting its size, so we have to count
-        // the elements by iterating over them...
-        let capacity = t.iter().count();
-        let mut extensions = BumpVec::with_capacity_in(capacity, self.bump);
-        let mut rest = None;
-
-        for ext in t.iter() {
-            if let Ok(index) = ext.parse::<usize>() {
-                // Extension sets in the model support at most one variable. This is a
-                // deliberate limitation so that extension sets behave like polymorphic rows.
-                // The type theory of such rows and how to apply them to model (co)effects
-                // is well understood.
-                //
-                // Extension sets in `hugr-core` at this point have no such restriction.
-                // However, it appears that so far we never actually use extension sets with
-                // multiple variables, except for extension sets that are generated through
-                // property testing.
-                if rest.is_some() {
-                    // TODO: We won't need this anymore once we have a core representation
-                    // that ensures that extension sets have at most one variable.
-                    panic!("Extension set with multiple variables")
-                }
+    pub fn export_ext_set(&mut self, ext_set: &ExtensionSet) -> model::TermId {
+        let capacity = ext_set.iter().size_hint().0;
```

Comment:
Hmmm, so `ExtensionSet::iter()` just returns an `impl Iterator`. I wasn't familiar with this `size_hint` but...are we sure `BumpVec::from_iter_in` doesn't use it?

Reply:
It fact it does - `from_iter_in` goes to `extend` [here](https://github.com/fitzgen/bumpalo/blob/8392c63d956962011a127bfd2f98c64aa36ff24a/src/collections/vec.rs#L2149-L2159)

Reply:
`size_hint` gives a lower and an optional upper bound for the contents of an iterator. Extension sets do not report their size with the current API, and I didn't want to do a `hugr-core` API change in this PR.

`BumpVec::from_iter_in` very likely uses the `size_hint`. Same reason here for `BumpVec::push` as in other places: this makes it a bit easier to refactor for now, when everything is still quite in flux.